### PR TITLE
Paypal marketing opt in

### PR DESCRIPTION
**Pull request by pvighi**
Added a post payment step when paying with PayPal that shows the option to sign in for marketing emails.

Some notes:
- The "post payment" page looks like the contribute page but it's a new page.All common parts have been extracted to a fragment named "contributionsMain" (worst name ever).
- To avoid any problems the post payment page will expect the email of the user to be in the play session.

### CODE REVIEW

**Comment by pvighi** [comment]
postpayment

File: app/controllers/PaypalController.scala [hunk 1]
```diff
@@ -30,17 +38,36 @@ class PaypalController(
...
+
     paypalService.executePayment(paymentId, payerId) match {
-      case Right(_) =>
+      case Right(executedPayment) =>
         paypalService.storeMetaData(paymentId, chosenVariants, cmp, intCmp, ophanId, idUser)
-        Redirect(routes.Giraffe.thanks(countryGroup).url, Map("CMP" -> cmp.toSeq, "INTCMP" -> intCmp.toSeq), SEE_OTHER)
+        getEmail(executedPayment) match {
+          case Some(email) => redirectWithCampaignCodes(postPayUrl).withSession(request.session + ("email" -> email))
+          case None =>
+            Logger.error("No email returned from Paypal payment execution, skipping to thank you page")
```
Comment (alexduf) [inline app/controllers/PaypalController.scala]: I think you should log the payment ID here

File: app/controllers/PaypalController.scala [hunk 2]
```diff
@@ -121,4 +149,27 @@ class PaypalController(
...
+  case class MetadataUpdate(marketingOptIn: Boolean)
+
+  val metadataUpdateForm: Form[MetadataUpdate] = Form(
+    mapping(
+      "marketingOptIn"->boolean
+    )(MetadataUpdate.apply)(MetadataUpdate.unapply)
+  )
+
+
+  def updateMetadata(countryGroup: CountryGroup) = NoCacheAction.async { implicit request =>
```
Comment (alexduf) [inline app/controllers/PaypalController.scala]: If you want to use a form I believe you can use `NoCacheAction(parse.form(metadataUpdateForm))` and then you don't have to do the `bindFromRequest`.

File: app/controllers/Redirect.scala [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package controllers
+
+import play.api.mvc.{Controller, Request}
+
+trait Redirect {self: Controller =>
```
Comment (alexduf) [inline app/controllers/Redirect.scala]: spacing! :fearful: 

`trait Redirect { self: Controller =>`
or

```
trait Redirect {
  self: Controller =>
```

File: app/controllers/Redirect.scala [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  def redirectWithCampaignCodes(destinationUrl: String, additionalParams: Set[String] = Set.empty)(implicit request: Request[Any]) = {
+    val QueryParamsToForward = Set("INTCMP", "CMP") ++ additionalParams
```
Comment (alexduf) [inline app/controllers/Redirect.scala]: `QueryParamsToForward` => `queryParamsToForward`

File: app/services/PaypalService.scala [hunk 5]
```diff
@@ -107,20 +107,24 @@ class PaypalService(config: PaypalApiConfig, contributionData: ContributionData)
...
+          Right(payment)
         }
-      case Failure(exception) => Left(exception.getMessage)
+      case Failure(exception) =>
+        Logger.error("Unable to execute payment", exception)
+        Left(exception.getMessage)
     }
   }
 
+  case class SavedContribution(contributionMetaData: ContributionMetaData, contributor: Contributor)
```
Comment (alexduf) [inline app/services/PaypalService.scala]: that should go into the `models` packages

File: assets/stylesheets/giraffe-react.scss [hunk 6]
```diff
@@ -8,7 +8,34 @@ $control-spacing: 5px;
...
+.contactInformationTitle {
+    background-image: url('/assets/images/inline-svgs/share-email.svg');
+    background-repeat: no-repeat;
+    background-size: 20px;
+    padding-left: 25px;
+
 
+}
+.postPaymentTick{
+    width:40px;
```
Comment (alexduf) [inline assets/stylesheets/giraffe-react.scss]: spacing :fearful:

**Review by alexduf — CHANGES_REQUESTED** [review]
Sorry I got a bit pedantic on this one but I'm afraid there's a couple of things to fix here

**Comment by pvighi** [comment]
new design:
newpostpay

**Review by alexduf — APPROVED** [review]

**Comment by alexduf** [comment]
:+1:

**Review by alexduf — APPROVED** [review]

**Pull request merged**